internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,202 @@
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package loki
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/cmd"
+	"github.com/canonical/pebble/internals/logger"
+	"github.com/canonical/pebble/internals/plan"
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+const maxRequestEntries = 100
+
+var requestTimeout = 10 * time.Second
+
+type Client struct {
+	targetName string
+	remoteURL  string
+	entries    []lokiEntryWithService
+
+	httpClient *http.Client
+}
+
+func NewClient(target *plan.LogTarget) *Client {
+	return &Client{
+		targetName: target.Name,
+		remoteURL:  target.Location,
+		httpClient: &http.Client{Timeout: requestTimeout},
+	}
+}
+
+func (c *Client) AddLog(entry servicelog.Entry) error {
+	c.entries = append(c.entries, lokiEntryWithService{
+		entry:   encodeEntry(entry),
+		service: entry.Service,
+	})
+	c.truncateBuffer()
+	return nil
+}
+
+func encodeEntry(entry servicelog.Entry) lokiEntry {
+	return lokiEntry{
+		strconv.FormatInt(entry.Time.UnixNano(), 10),
+		strings.TrimSuffix(entry.Message, "\n"),
+	}
+}
+
+func (c *Client) NumBuffered() int {
+	return len(c.entries)
+}
+
+func (c *Client) Flush(ctx context.Context) error {
+	if c.NumBuffered() == 0 {
+		return nil // no-op
+	}
+
+	req := c.buildRequest()
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+// resetBuffer drops all buffered logs (in the case of a successful send, or an
+// unrecoverable error).
+func (c *Client) resetBuffer() {
+	c.entries = c.entries[:0]
+}
+
+// truncateBuffer truncates the buffer to maxRequestEntries, removing the
+// oldest logs first.
+func (c *Client) truncateBuffer() {
+	l := len(c.entries)
+	if l > maxRequestEntries {
+		c.entries = c.entries[(l - maxRequestEntries):]
+	}
+}
+
+func (c *Client) buildRequest() lokiRequest {
+	// Put entries into service "buckets"
+	bucketedEntries := map[string][]lokiEntry{}
```

Comment:
Maybe it doesn't matter, but the nice thing about how you were doing it before was that we didn't have to reallocate the map and its slices every request -- significantly fewer allocations per log and per request. But probably the overhead of the request, at least for small requests, is so much bigger that it doesn't matter. So I've convinced myself it's okay as is, at least to start with. :-)

Reply:
the reason I moved back to a slice is the new truncation logic. It would be very hard to "truncate to the last 100 entries" if we were still indexing them by service - we'd have to keep a timestamp for when each one was added, and iterate over all of them, etc.